### PR TITLE
Add semantic conventions for GCP resources

### PR DESCRIPTION
**Pull request by damemi**
Fixes #

## Changes

This adds GCP-specific semantic conventions for GCE instances. Specifically, we would like to provide 2 ways of identifying host names: by their GCE instance name and their fully-qualified hostname (see https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/issues/511 for more downstream discussion)

(Following the pattern for AWS in https://github.com/open-telemetry/opentelemetry-specification/pull/1099)

Related issues # https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/issues/511

Related [OTEP(s)](https://github.com/open-telemetry/oteps) #

### CODE REVIEW

**Comment by damemi** [comment]
cc @punya @jsuereth @aabmass @psx95

**Comment by punya** [comment]
Do all GCE instances have a well-defined and unique fully qualified hostname?

**Comment by damemi** [comment]
> Do all GCE instances have a well-defined and unique fully qualified hostname?

I think so, based on https://cloud.google.com/compute/docs/instances/custom-hostname-vm:

> When you create a virtual machine (VM) instance, Google Cloud [creates an internal DNS name](https://cloud.google.com/compute/docs/internal-dns#about_internal_dns) from the VM name. Unless you specify a custom hostname, Google Cloud uses the automatically created internal DNS name as the hostname it provides to the VM.

**Comment by punya** [comment]
Thanks for the documentation pointer. It looks like the internal DNS name is always defined, and (for projects created since 2018), has the form `VM_NAME.ZONE.c.PROJECT_ID.internal`, which is redundant with information we put into other resource attributes already. Can we learn more about the benefit/goal of adding this redundant information? If it's only useful to a small subset of users, might it be better for them to construct it (e.g. using a processor) rather than taking up another attribute from the (often limited) limit of attributes for everyone?

**Comment by damemi** [comment]
@punya is that just the default DNS name? There could still be a benefit when setting a custom hostname that differs from that default. In that case `instanceName != hostName != DNS name`

Documenting how to get the default would be useful too. Do you have a link to where you found that?

**Comment by punya** [comment]
> Documenting how to get the default would be useful too. Do you have a link to where you found that?

It's a link from your previous comment, https://cloud.google.com/compute/docs/internal-dns

**Comment by damemi** [comment]
@punya ah I missed that definition (I just copied the higher level explanation from the custom hostnames doc)

Do you think custom hostnames are a reasonable justification for this?

**Comment by mx-psi** [comment]
> An alternative could be to make this a config option in the resource detector. I don't know what level of config we can provide in that processor, but something like hostnameValue: <instance/hostname> could also give users the choice.

I think that defeats the purpose of semantic conventions; they should have a single well-defined meaning so that backends can reliably interpret them as one thing or the other. If `host.name` can have 5 different meanings, we either have broken setups or have to rely on other values (something like `hostnameValue.config: instance`, which I think is not a good idea)

**Comment by punya** [comment]
Should we consider breaking backwards compatibility in the GCP resource detector, and stop setting `host.name`?

**Comment by damemi** [comment]
@punya I don't think there is a need to break the existing convention right away. Could we recommend users switch to one of the new attributes and plan to stop setting `host.name` at some point in the future?

It is a bit weird to set the Instance name in 2 places (`host.name` and `gcp.gce.instance.name`), but the new attribute is more rigidly defined ie "This is always exactly the GCP GCE instance name" vs the current definition for the generic attribute:

> Name of the host. On Unix systems, it may contain what the hostname command returns, or the fully qualified hostname, or another name specified by the user.

**Comment by punya** [comment]
My apologies, I'll be out of the office for the next 2 weeks. @jsuereth has graciously agreed to handle the remaining review work.

File: semantic_conventions/resource/cloud_provider/gcp/gce.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    type: resource
+    brief: >
+      Resources used by Google Compute Engine (GCE).
+    attributes:
+      - id: instance.name
+        type: string
+        brief: >
+          The instance name of a GCE instance.
+        examples: ['instance-1']
+      - id: instance.hostname
```
Comment (jsuereth) [inline semantic_conventions/resource/cloud_provider/gcp/gce.yaml:13]: Should this just `ref: host.name` instead of define a new field?

Comment (damemi) [reply semantic_conventions/resource/cloud_provider/gcp/gce.yaml:13]: Yeah that sounds good, but `gcp.gce.instance.name` would actually be the new field that should `ref: host.name` (currently `host.name` is set to the instance name)

Comment (damemi) [reply semantic_conventions/resource/cloud_provider/gcp/gce.yaml:13]: Confirming what we discussed, we use [the *instance* name](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/327f5866b4fc60227d44c666c9ccb160a8ef9bf6/detectors/gcp/gce.go#L42-L44) to [set `host.name`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/97b4d5c41753b58a2273bb27259160f963e3cd8b/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L102) (also for [GKE](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/97b4d5c41753b58a2273bb27259160f963e3cd8b/processor/resourcedetectionprocessor/internal/gcp/gcp.go#L71)). We don't currently set the actual hostname anywhere.

**Comment by damemi** [comment]
@jsuereth I added much more description to these fields. I think the original decision to use Instance name as the `host.name` (ie the name of the host) was reasonable in hindsight, as most users probably just want the VM name rather than the full `my-vm.us-east.project.c.internal` default DNS name. It is confusingly overloaded with the actual term `hostname` (ie the hostname of the host), so I'm fine switching that if you think it's better.

**Comment by reyang** [comment]
@damemi heads up - most likely this PR will be closed, and we'll ask you to resubmit the PR in a new repo, please refer to https://github.com/open-telemetry/opentelemetry-specification/pull/3474#issuecomment-1539386505.

**Review by jsuereth — CHANGES_REQUESTED** [review]
Can you move this to https://github.com/open-telemetry/semantic-conventions?

**Comment by damemi** [comment]
> Can you move this to https://github.com/open-telemetry/semantic-conventions?

Done, https://github.com/open-telemetry/semantic-conventions/pull/15

**Pull request closed without merging**